protocol/x/govplus/keeper/msg_server.go

```diff
@@ -15,3 +21,40 @@ func NewMsgServerImpl(keeper Keeper) types.MsgServer {
 }
 
 var _ types.MsgServer = msgServer{}
+
+func (k msgServer) SlashValidator(
+	goCtx context.Context,
+	msg *types.MsgSlashValidator,
+) (*types.MsgSlashValidatorResponse, error) {
+	if !k.HasAuthority(msg.Authority) {
+		return nil, errorsmod.Wrapf(
+			govtypes.ErrInvalidSigner,
+			"invalid authority %s",
+			msg.Authority,
+		)
+	}
+
+	consAddr, err := sdk.ConsAddressFromBech32(msg.ValidatorAddress)
+	if err != nil {
+		return nil, types.ErrValidatorAddress
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	_, err = k.stakingKeeper.Slash(
+		ctx,
+		consAddr,
+		int64(msg.InfractionHeight),
+		msg.PowerAtInfractionHeight.BigInt().Int64(),
```

Comment:
It seems pretty bareboned that proposer would have to input the voting power themselves. Can we at least calculating the voting power from the delegation coin amount?

From `Slash()` [implementation](https://github.com/cosmos/cosmos-sdk/blob/b947b02dc63db0a8011a96e217197bf4559dea92/x/staking/keeper/slash.go#L46-L47), it calculates the amount to check with no additional validation on this figure. If proposer mistakenly put a larger than actual amount, would this lead to any delegator being slashed more than its delegation amount? In general, it seems like we need some E2Es test to cover possible scenarios.

Reply:
The `Slash()` function in `x/staking` was not meant to be exposed for usage by human, and it has a couple of input [contracts](https://github.com/cosmos/cosmos-sdk/blob/b947b02dc63db0a8011a96e217197bf4559dea92/x/staking/keeper/slash.go#L20-L36) listed. Can we explicitly check these contracts (if not checked within the function) to make sure there are possible weird cases?

Reply:
You cannot because you can only get the delegation coin amount at the time of execution. If someone sees a proposal to slash them and knows they can avoid it by manipulating their delegation amount, it becomes worthless

Reply:
> Can we at least calculating the voting power from the delegation coin amount?
I meant having user input the total delegated coin amount, as opposed to voting power. The conversion is [trivial](https://github.com/cosmos/cosmos-sdk/blob/b947b02dc63db0a8011a96e217197bf4559dea92/types/staking.go#L35) and can just be done on protocol.

Reply:
Just [in genesis state.](https://github.com/dydxprotocol/cosmos-sdk/blob/dydx-fork-v0.50.3/proto/cosmos/staking/v1beta1/genesis.proto#L27). I can change this to tokens instead if you still prefer it that way. Let me know

Reply:
> Can we explicitly check these contracts (if not checked within the function)?
All parts of the contract are checked in the Slash function and will error out. I've added to ValidateBasic as well.

(Except "Infraction was committed equal to or less than an unbonding period in the past, so all unbonding delegations and redelegations from that height are stored". It is up to the community to slash before the tokens are gone.)